x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/inference/nlp/tokenizers/BpeTokenizer.java

```diff
@@ -0,0 +1,447 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+package org.elasticsearch.xpack.ml.inference.nlp.tokenizers;
+
+import com.carrotsearch.hppc.CharObjectHashMap;
+import com.carrotsearch.hppc.CharObjectMap;
+
+import org.apache.lucene.analysis.CharArrayMap;
+import org.apache.lucene.analysis.CharArraySet;
+import org.apache.lucene.analysis.Tokenizer;
+import org.apache.lucene.analysis.tokenattributes.CharTermAttribute;
+import org.apache.lucene.analysis.tokenattributes.OffsetAttribute;
+import org.apache.lucene.analysis.tokenattributes.PositionIncrementAttribute;
+import org.elasticsearch.common.Strings;
+
+import java.io.IOException;
+import java.io.Reader;
+import java.nio.charset.StandardCharsets;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Map;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+import java.util.stream.Collectors;
+import java.util.stream.IntStream;
+
+public class BpeTokenizer extends Tokenizer {
+    private final CharTermAttribute termAtt = addAttribute(CharTermAttribute.class);
+    private final OffsetAttribute offsetAtt = addAttribute(OffsetAttribute.class);
+    private final PositionIncrementAttribute posIncAtt = addAttribute(PositionIncrementAttribute.class);
+
+    private static final Pattern pattern = Pattern.compile(
+        "'s|'t|'re|'ve|'m|'ll|'d| ?\\p{L}+| ?\\p{N}+| ?[^\\s\\p{L}\\p{N}]+|\\s+(?!\\S)|\\s+"
+    );
+
+    static Map<Integer, Character> byteEncoder() {
+        List<Integer> bytes = IntStream.concat(
+            IntStream.range(Character.codePointAt("!", 0), Character.codePointAt("~", 0) + 1),
+            IntStream.concat(
+                IntStream.range(Character.codePointAt("¡", 0), Character.codePointAt("¬", 0) + 1),
+                IntStream.range(Character.codePointAt("®", 0), Character.codePointAt("ÿ", 0) + 1)
+            )
+        ).boxed().collect(Collectors.toList());
+        List<Integer> chars = new ArrayList<>(bytes);
+        int n = 0;
+        for (int i = 0; i < 2 << 7; i++) {
```

Comment:
Is `2 << 7` really clearer than `256`?  There are other places in this PR where `256` is hardcoded, which relies on the reader knowing that there are `256` possible values in a byte.  If we expect the reader to know that then I don't see why `256` cannot be used in this method too.